### PR TITLE
SegmentTree to Data Structures

### PR DESCRIPTION
**Pull request by forgotter**
Segment Tree is a very useful data structure which helps perform various operation in logarithmic time. A bottom-up implementation of segment tree is done and it answers queries of sum, min, max over a range.

### CODE REVIEW

**Comment by juanplopes** [comment]
Very elegant implementation.

**Comment by forgotter** [comment]
Added <b>changePriority</b> feature in priority queue, it works in <i>amortized constant time</i>. The priority queue implementation is done from scratch.

**Comment by juanplopes** [comment]
@forgotter I think it would be better to push this change into another branch, and open a new pull request just for it.

**Comment by forgotter** [comment]
Yeah, I should have done it that way. Thanks :).

**Comment by forgotter** [comment]
@juanplopes Hey, Is there a way to revert back to previous point. Or should I delete the fork and re-fork and then create branches for individual additions and create a pull request for each of them? Thanks for your help, I am just getting started with Github.

File: src/data_structures/segment_tree.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      this._elementsMin[i] = 0;
+      this._elementsMax[i] = 0;
+    }
+  }
+
+  /**
+  Updates the value at index with the given value
+  */
+  modify(index, value) {
+    index = index + this._size;
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: Don't reassign values to passed parameters

File: src/data_structures/segment_tree.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      this._elementsMax[i >> 1] = Math.max(this._elementsMax[i], this._elementsMax[i ^ 1]);
+    }
+  }
+
+  /**
+  Return the sum of all elements in the range [from, to]
+  */
+  getSum(from, to) {
+    let ans = 0;
+    for (from += this._size, to += this._size + 1; from < to; from >>= 1, to >>= 1) {
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: Don't reassign values to passed parameters

File: src/data_structures/segment_tree.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      this._elementsMin[i >> 1] = Math.min(this._elementsMin[i], this._elementsMin[i ^ 1]);
+      this._elementsMax[i >> 1] = Math.max(this._elementsMax[i], this._elementsMax[i ^ 1]);
+    }
+  }
+
+  /**
+  Return the sum of all elements in the range [from, to]
+  */
+  getSum(from, to) {
+    let ans = 0;
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: What's `ans`? Can this variable be named better?

File: src/data_structures/segment_tree.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      if (to % 2 == 1) ans += this._elementsSum[--to];
+    }
+    return ans;
+  }
+
+  /**
+  Return the minimum of all elements in the range [from, to]
+  */
+  getMin(from, to) {
+    let ans = Number.MAX_SAFE_INTEGER;
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: What's `ans`? Can this variable be named better?

File: src/data_structures/segment_tree.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    for (from += this._size, to += this._size + 1; from < to; from >>= 1, to >>= 1) {
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: Don't reassign values to passed parameters

File: src/data_structures/segment_tree.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      if (to % 2 == 1) ans = Math.min(ans, this._elementsMin[--to]);
+    }
+    return ans;
+  }
+
+  /**
+  Return the maximum of all elements in the range [from, to]
+  */
+  getMax(from, to) {
+    let ans = Number.MIN_SAFE_INTEGER;
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: What's `ans`? Can this variable be named better?

File: src/data_structures/segment_tree.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    for (from += this._size, to += this._size + 1; from < to; from >>= 1, to >>= 1) {
```
Comment (felipernb) [inline src/data_structures/segment_tree.js]: Don't reassign values to passed parameters

**Comment by forgotter** [comment]
Sure, I'll try to refactor the code with function and operator.  

Make sure not to reassign values to passed parameter. 

About ans variable, it's just the answer for the given function in the asked range, i mean totalSum while summing, minValue while minimum over range and maxValue while maximum over range.
 
Yeah, forgot to add test file. Will add in next PR.

P.S. Should I work in linux to get rid of CLRS issue?

**Comment by forgotter** [comment]
I tried to merge getMin, getMax and getSum but seems like it can't be done. Since each of them performs a different function on different tree, the code can't be reduced but will only increase the overhead by a function call.

Actually, most of time, we are only interested in performing one of functions in entire range, i.e. either min or max or sum. Maybe something like implement a segment tree class and then extending it to SumSegmentTree, MinSegmentTree and MaxSegementTree can be done.

**Comment by juanplopes** [comment]
Perhaps you can do something like that (didn't test it, may contain syntax errors):

```javascript
  aggregate(from, to, where, initialValue, aggregateFn) {
    let ans = initialValue;
    for (from += this._size, to += this._size + 1; from < to; from >>= 1, to >>= 1) {
      if (from % 2 == 1) ans = aggregateFn(ans, where[from++]);
      if (to % 2 == 1) ans = aggregateFn(ans, where[--from]);
    }
    return ans;
  }

  /**
  Return the sum of all elements in the range [from, to]
  */
  getSum(from, to) {
    return aggregate(from, to, this._elementsSum, 0, (a, b) => a + b);
  }

  /**
  Return the minimum of all elements in the range [from, to]
  */
  getMin(from, to) {
    return aggregate(from, to, this._elementsMin, Number.MAX_SAFE_INTEGER, Math.min);
  }

  /**
  Return the maximum of all elements in the range [from, to]
  */
  getMax(from, to) {
    return aggregate(from, to, this._elementsMax, Number.MIN_SAFE_INTEGER, Math.max);
  }
 
```

**Comment by forgotter** [comment]
Quite an implementation I'd say. Perfect use of lambda functions.  There's only one typo where[--to] instead of where[--from], Thanks for code.

What do you think of different implementation of each of individual function, separate classes for each of Min, Max and Sum. Since mostly only one of function needs to be performed on a range.  In current implementation, the complexity of modify is 3*(log(n)), by implementing it in form of separate classes, it'd be around (log(n)) and also only one array would be required for storage compared to three as with current implementation.

When I implementing it initially, my only thoughts were to put users to ease while using it, but now I think that It'd be better to do it in form of separate classes.

**Comment by forgotter** [comment]
The Travis-CI build fails pointing no such path exists. What should be the path in the test-file?

**Comment by forgotter** [comment]
@felipernb The current implementation looks fine to me. I have tried to include all suggested changes, maybe add more comments explaining function if required.

Everything seems to be working. sonar-scanner fails in Travis-CI, please check it out.

**Comment by felipernb** [comment]
Thanks for the contribution and sorry for the delay. Merging now!

**Pull request merged**